### PR TITLE
Conf ver

### PR DESCRIPTION
**Pull request by eevans**
Adds a version attribute to the module configuration, and persists it to the meta.value schema structure as a hash-excluded private attribute (`_config_version`).  Replication changes are conditional on a monotonically increasing configuration version.

Bug: https://phabricator.wikimedia.org/T76494

### CODE REVIEW

**Comment by Pchelolo** [comment]
From a code perspective this looks good to me, however I do agree with a point you're making in the ticket - this is a second step on a road to special-cases-hell. 

As I understand, the `_backend_version` and `_config_version` doesn't belong to a table schema, it's an application-vide metadata, so in the perfect world we could have a special `app.meta` table to hold it in a single place. I don't propose to do such a drastic change now, but let's consider it if we add yet another exceptional case.

Overall, LGTM.

**Comment by gwicke** [comment]
@Pchelolo, backend version migrations need to be tracked per-table, so a global table wouldn't work.

**Comment by Pchelolo** [comment]
@gwicke actually, I don't understand why. If I get all the code correctly, after each deploy each table will have equal backend version (right now it's 1), and on `CURRENT_BACKEND_VERSION` increase, each table will go through the backend migration, and again, have equal value.

**Comment by eevans** [comment]
@Pchelolo 
> @gwicke actually, I don't understand why. If I get all the code correctly, after each deploy each table will have equal backend version (right now it's 1), and on CURRENT_BACKEND_VERSION increase, each table will go through the backend migration, and again, have equal value.

Backend migrations are applied on a table-by-table basis though, and the `_backend_version` attribute is only updated after a successful migration.

FWIW, my suggestion was to move `_{backend,config}_version` out of the `meta.value` JSON blob, to separate, but still per-table attributes.

**Comment by Pchelolo** [comment]
@eevans ah, ok, my suggestion was indeed not good.

Anyway, the current code looks good to me.

**Comment by gwicke** [comment]
Another option we could consider is including all those versions in the quick-check hash. We are interested in speeding up the common case of startup with nothing having changed. In that case, only checking the hash will be slightly more efficient and possibly less messy than checking the versions separately.

The only downside I see in doing it that way is the possibility of a small slow-down when only the backend version or config version has changed, as all parts of the schema will still be checked for changes. I would be surprised if this difference mattered though, compared to the hundreds / thousands of checks without hash changes.

**Comment by eevans** [comment]
@gwicke 
> Another option we could consider is including all those versions in the quick-check hash. We are interested in speeding up the common case of startup with nothing having changed. In that case, only checking the hash will be slightly more efficient and possibly less messy than special-casing the version checks.

That's basically the approach I was referring to [here](https://phabricator.wikimedia.org/T76494#1740006).  The problem is that for the table schema parts, you'll still need to compare the actual content of the persisted versus requested schema (sans the other two versions).  It could work, it just really doesn't feel like much of a win when weighing the additional complexity against optimizing away a couple of integer comparisons in the check.  YMMV though.

File: lib/dbutils.js [hunk 1]
```diff
@@ -215,15 +218,20 @@ dbu.eachRow = function eachRow(client, query, params, options, handler) {
...
  */
-dbu.validateAndNormalizeSchema = function validateAndNormalizeSchema(schema) {
+dbu.validateAndNormalizeSchema = function validateAndNormalizeSchema(schema, configVer) {
     if (!schema._backend_version) {
         schema._backend_version = dbu.CURRENT_BACKEND_VERSION;
     }
+    if (configVer) {
+        schema._config_version = configVer;
+    }
     return validator.validateAndNormalizeSchema(schema);
```
Comment (gwicke) [inline lib/dbutils.js]: I'm actually wondering if we should add both `_backend_version` and `_schema_version` after `validator.validateAndNormalizeSchema(schema)`, to make sure that they don't go missing. We have a todo in the schema validator to validate other attributes (https://github.com/wikimedia/restbase-mod-table-spec/blob/master/lib/validator.js#L199), which could lead to that result.

**Comment by gwicke** [comment]
See #166 for a rebased & refactored version of this.

**Comment by gwicke** [comment]
Subsumed by #166.

**Pull request closed without merging**